File: elixir/ltx/test/security_test.exs
# test/security_test.exs — Epic 29 security tests for InterplanetLtx (Elixir)
# Run with: elixir -r test/test_helper.exs test/security_test.exs

Code.require_file("../lib/interplanet_ltx/constants.ex", __DIR__)
Code.require_file("../lib/interplanet_ltx/models.ex", __DIR__)
Code.require_file("../lib/interplanet_ltx/interplanet_ltx.ex", __DIR__)
Code.require_file("../lib/interplanet_ltx/security.ex", __DIR__)

import Test

alias InterplanetLtx.Security

# ── 1. canonical_json_key_order ───────────────────────────────────────────────

a = Security.canonical_json(%{"b" => 1, "a" => 2})
b = Security.canonical_json(%{"a" => 2, "b" => 1})
check a == b,                       "canonical_json: same output regardless of insertion order"
check a == "{\"a\":2,\"b\":1}",     "canonical_json: keys sorted, correct format"

check Security.canonical_json(42)   == "42",    "canonical_json integer"
check Security.canonical_json("hi") == "\"hi\"","canonical_json string"
check Security.canonical_json(true) == "true",  "canonical_json true"
check Security.canonical_json(nil)  == "null",  "canonical_json null"
check Security.canonical_json([1, 2]) == "[1,2]","canonical_json array"

nested = Security.canonical_json(%{"z" => %{"b" => 1, "a" => 2}})
check nested == "{\"z\":{\"a\":2,\"b\":1}}", "canonical_json nested map sorted"

# ── 2. generate_nik_fields ────────────────────────────────────────────────────

result = Security.generate_nik(valid_days: 365, node_label: "test-node")
nik    = result.nik

check is_binary(nik["nodeId"]),               "generate_nik: nodeId is string"
check String.length(nik["nodeId"]) > 0,       "generate_nik: nodeId non-empty"
check is_binary(nik["publicKey"]),            "generate_nik: publicKey is string"
check String.length(nik["publicKey"]) > 0,   "generate_nik: publicKey non-empty"
check nik["algorithm"] == "Ed25519",          "generate_nik: algorithm is Ed25519"
check nik["keyVersion"] == 1,                 "generate_nik: keyVersion is 1"
check nik["label"] == "test-node",            "generate_nik: label stored"
check is_binary(result.private_key_b64),      "generate_nik: private_key_b64 is string"
check String.length(result.private_key_b64) > 0, "generate_nik: private_key_b64 non-empty"

# Two NIKs should have different node IDs
result2 = Security.generate_nik()
check result.nik["nodeId"] != result2.nik["nodeId"], "generate_nik: unique nodeIds"

# ── 3. is_nik_expired_false ───────────────────────────────────────────────────

future_nik = %{"validUntil" => "2099-12-31T23:59:59.000Z"}
check Security.is_nik_expired(future_nik) == false, "is_nik_expired: future date -> false"

fresh_result = Security.generate_nik(valid_days: 365)
check Security.is_nik_expired(fresh_result.nik) == false, "is_nik_expired: fresh NIK not expired"

# ── 4. is_nik_expired_true ────────────────────────────────────────────────────

past_nik = %{"validUntil" => "2000-01-01T00:00:00.000Z"}
check Security.is_nik_expired(past_nik) == true, "is_nik_expired: past date -> true"

expired_nik = %{"validUntil" => "1970-01-01T00:00:00.000Z"}
check Security.is_nik_expired(expired_nik) == true, "is_nik_expired: epoch -> true"

# ── 5. sign_verify_roundtrip ──────────────────────────────────────────────────

nik_result = Security.generate_nik(valid_days: 365)
test_plan  = %{"v" => 2, "title" => "Test Session", "start" => "2099-01-01T00:00:00Z"}

# Pass the pub_key_b64 so kid can be derived correctly
signed = Security.sign_plan(test_plan, nik_result.private_key_b64, nik_result.nik["publicKey"])

check is_map(signed), "sign_plan: returns map"
check is_map(signed.coseSign1), "sign_plan: coseSign1 present"
check is_binary(signed.coseSign1["signature"]), "sign_plan: signature is string"
check is_binary(signed.coseSign1["payload"]),   "sign_plan: payload is string"
check is_binary(signed.coseSign1["protected"]), "sign_plan: protected is string"

# The kid should match the nodeId in the NIK
kid = signed.coseSign1["unprotected"]["kid"]
check kid == nik_result.nik["nodeId"], "sign_plan: kid matches nodeId"

key_cache = %{nik_result.nik["nodeId"] => nik_result.nik}
result_v  = Security.verify_plan(signed, key_cache)

check result_v.valid == true,             "verify_plan: roundtrip valid"
check Map.get(result_v, :reason) == nil,  "verify_plan: no reason on success"

# ── 6. sign_verify_tampered ───────────────────────────────────────────────────

nik_result2 = Security.generate_nik(valid_days: 365)
test_plan2  = %{"v" => 2, "title" => "Original Plan", "start" => "2099-01-01T00:00:00Z"}

signed2  = Security.sign_plan(test_plan2, nik_result2.private_key_b64, nik_result2.nik["publicKey"])
tampered = Map.put(signed2, :plan, %{"v" => 2, "title" => "TAMPERED!", "start" => "2099-01-01T00:00:00Z"})

key_cache2 = %{nik_result2.nik["nodeId"] => nik_result2.nik}
result_t   = Security.verify_plan(tampered, key_cache2)

check result_t.valid == false,              "verify_plan: tampered plan invalid"
check result_t.reason == "payload_mismatch","verify_plan: reason=payload_mismatch"

# ── 7. sign_verify_wrong_key ──────────────────────────────────────────────────

nik_result3 = Security.generate_nik(valid_days: 365)
test_plan3  = %{"v" => 2, "title" => "Plan", "start" => "2099-01-01T00:00:00Z"}
signed3     = Security.sign_plan(test_plan3, nik_result3.private_key_b64, nik_result3.nik["publicKey"])

# Empty key cache
result_wk = Security.verify_plan(signed3, %{})
check result_wk.valid == false,            "verify_plan: empty cache invalid"
check result_wk.reason == "key_not_in_cache","verify_plan: reason=key_not_in_cache"

# Wrong kid in cache (different node's NIK keyed by signer's nodeId)
wrong_nik    = Security.generate_nik(valid_days: 365)
wrong_cache  = %{nik_result3.nik["nodeId"] => wrong_nik.nik}
result_wk2   = Security.verify_plan(signed3, wrong_cache)
check result_wk2.valid == false, "verify_plan: wrong key in cache invalid"

# ── 8. sequence_tracker_replay ────────────────────────────────────────────────

tracker = Security.new_sequence_tracker("plan-001")

b1 = Security.add_seq(%{data: "hello"}, tracker, "N0")
check b1[:seq] == 1, "add_seq: first seq is 1"

b2 = Security.add_seq(%{data: "world"}, tracker, "N0")
check b2[:seq] == 2, "add_seq: second seq is 2"

# Receive b1 (seq=1) — accepted
r1 = Security.check_seq(b1, tracker, "N0")
check r1.accepted == true, "check_seq: first message accepted"
check r1.gap == false,     "check_seq: no gap on seq 1"
check r1.gap_size == 0,    "check_seq: gap_size 0"

# Replay b1 — rejected
r2 = Security.check_seq(b1, tracker, "N0")
check r2.accepted == false,  "check_seq: replay rejected"
check r2.reason == "replay", "check_seq: reason=replay"

# b2 in order
r3 = Security.check_seq(b2, tracker, "N0")
check r3.accepted == true,  "check_seq: seq 2 accepted"
check r3.gap == false,      "check_seq: no gap"

# seq=5 — gap of 2
b5 = %{data: "skip", seq: 5}
r4 = Security.check_seq(b5, tracker, "N0")
check r4.accepted == true, "check_seq: gap accepted"
check r4.gap == true,      "check_seq: gap detected"
check r4.gap_size == 2,    "check_seq: gap_size=2"

# N1 independent counter
b1_n1 = Security.add_seq(%{data: "from N1"}, tracker, "N1")
check b1_n1[:seq] == 1, "add_seq: N1 starts at 1"

# ── 9. key_expired ────────────────────────────────────────────────────────────

nik_r4    = Security.generate_nik(valid_days: 365)
test_p4   = %{"v" => 2, "title" => "P4", "start" => "2099-01-01T00:00:00Z"}
signed4   = Security.sign_plan(test_p4, nik_r4.private_key_b64, nik_r4.nik["publicKey"])

expired2    = Map.put(nik_r4.nik, "validUntil", "2000-01-01T00:00:00.000Z")
exp_cache   = %{nik_r4.nik["nodeId"] => expired2}
result_exp  = Security.verify_plan(signed4, exp_cache)
check result_exp.valid == false,           "verify_plan: expired key invalid"
check result_exp.reason == "key_expired",  "verify_plan: reason=key_expired"

# ── Summary ───────────────────────────────────────────────────────────────────

passed = Process.get(:passed, 0)
failed = Process.get(:failed, 0)
IO.puts("\n#{passed} passed  #{failed} failed")
if failed > 0, do: System.halt(1)

File: lua/ltx/test/security_test.lua
-- security_test.lua -- Epic 29, Stories 29.1 / 29.4 / 29.5

package.path = package.path .. ';../?.lua;../src/?.lua;./src/?.lua;./?.lua'
local SEC = require('security')

local passed = 0
local failed = 0

local function ok(cond, msg)
  if cond then
    passed = passed + 1
  else
    failed = failed + 1
    print('FAIL: ' .. tostring(msg))
  end
end

local function eq(a, b, msg)
  ok(a == b, (msg or '') .. '  expected=' .. tostring(b) .. '  got=' .. tostring(a))
end

-- ---- Story 29.1: canonical_json ----

eq(SEC.canonical_json({}), '{}', 'empty object')
eq(SEC.canonical_json({z=1, a=2}), '{"a":2,"z":1}', 'sorted keys')
eq(SEC.canonical_json({1,2,3}), '[1,2,3]', 'array')
eq(SEC.canonical_json(42), '42', 'number')
eq(SEC.canonical_json(true), 'true', 'boolean')
eq(SEC.canonical_json('hi'), '"hi"', 'string')
eq(SEC.canonical_json(nil), 'null', 'nil->null')

-- nested object
local nested = SEC.canonical_json({b={y=9,x=1}, a=3})
ok(nested == '{"a":3,"b":{"x":1,"y":9}}', 'nested sorted keys: ' .. nested)

-- ---- Story 29.1: generate_nik ----

local nik = SEC.generate_nik()
ok(nik ~= nil, 'generate_nik returns table')
eq(nik.key_type, 'ltx-nik-v1', 'key_type')
ok(nik.node_id ~= nil and #nik.node_id > 0, 'node_id non-empty')
ok(nik.kid ~= nil and #nik.kid > 0, 'kid non-empty')
ok(nik.public_key_b64 ~= nil and #nik.public_key_b64 > 0, 'pub key non-empty')
ok(nik.private_key_b64 ~= nil and #nik.private_key_b64 > 0, 'priv key non-empty')
ok(nik.issued_at ~= nil, 'issued_at set')
ok(nik.expires_at ~= nil, 'expires_at set')

-- node_id length: 16 raw bytes -> 22 base64url chars
ok(#nik.node_id == 22, 'node_id is 22 chars (16 bytes b64u): ' .. #nik.node_id)

-- custom label and validity
local nik2 = SEC.generate_nik({valid_days=30, node_label='TestNode'})
eq(nik2.node_label, 'TestNode', 'node_label preserved')
ok(nik2.expires_at > nik2.issued_at, 'expires_at after issued_at')

-- two NIKs must have different node_ids
local nik3 = SEC.generate_nik()
ok(nik.node_id ~= nik3.node_id, 'unique node_ids')

-- ---- Story 29.1: is_nik_expired ----

ok(not SEC.is_nik_expired(nik), 'fresh nik not expired')

local expired_nik = {expires_at = '2000-01-01T00:00:00Z'}
ok(SEC.is_nik_expired(expired_nik), 'old nik is expired')

ok(SEC.is_nik_expired(nil), 'nil nik is expired')
ok(SEC.is_nik_expired({}), 'nik with no expires_at is expired')

-- ---- Story 29.4: sign_plan / verify_plan ----

local plan = {planId='p1', startAt='2026-05-01T00:00:00Z', quantum=60}
local sp = SEC.sign_plan(plan, nik.private_key_b64, nik.public_key_b64, nik.kid)
ok(sp ~= nil, 'sign_plan returns table')
ok(sp.coseSign1 ~= nil, 'coseSign1 present')
ok(sp.coseSign1.protected ~= nil, 'protected header present')
ok(sp.coseSign1.signature ~= nil, 'signature present')
eq(sp.coseSign1.unprotected.kid, nik.kid, 'kid in unprotected header')

-- verify with matching key cache
local cache = {}
cache[nik.kid] = nik
local ok1, reason1 = SEC.verify_plan(sp, cache)
eq(ok1, true, 'verify ok')
eq(reason1, 'ok', 'reason is ok')

-- verify fails: key not in cache
local ok2, reason2 = SEC.verify_plan(sp, {})
eq(ok2, false, 'verify fails empty cache')
eq(reason2, 'key_not_in_cache', 'reason key_not_in_cache')

-- verify fails: key expired
local old_cache = {}
old_cache[nik.kid] = {expires_at = '2000-01-01T00:00:00Z', private_key_b64 = nik.private_key_b64}
local ok3, reason3 = SEC.verify_plan(sp, old_cache)
eq(ok3, false, 'verify fails expired key')
eq(reason3, 'key_expired', 'reason key_expired')

-- verify fails: payload mismatch
local tampered = {
  plan = {planId='TAMPERED', startAt='2026-05-01T00:00:00Z', quantum=60},
  coseSign1 = sp.coseSign1,
}
local ok4, reason4 = SEC.verify_plan(tampered, cache)
eq(ok4, false, 'verify fails tampered plan')
eq(reason4, 'payload_mismatch', 'reason payload_mismatch')

-- sign_plan with missing args returns nil
local sp_nil, err_nil = SEC.sign_plan(nil, nik.private_key_b64)
eq(sp_nil, nil, 'sign_plan nil plan returns nil')

-- ---- Story 29.5: SequenceTracker ----

local st = SEC.new_sequence_tracker('plan-x')
ok(st ~= nil, 'tracker created')
eq(st.plan_id, 'plan-x', 'plan_id stored')

-- first seq is always accepted
local r1, m1 = SEC.add_seq(st, 'alice', 1)
eq(r1, true, 'first seq accepted')
eq(m1, 'ok', 'first seq message ok')

-- next seq accepted
local r2, m2 = SEC.add_seq(st, 'alice', 2)
eq(r2, true, 'seq 2 accepted')
eq(m2, 'ok', 'seq 2 message ok')

-- replay (same seq)
local r3, m3 = SEC.add_seq(st, 'alice', 2)
eq(r3, false, 'replay rejected')
eq(m3, 'replay', 'replay reason')

-- gap
local r4, m4 = SEC.add_seq(st, 'alice', 10)
eq(r4, true, 'gap accepted')
eq(m4, 'gap', 'gap reason')

-- check_seq (read-only)
local c1, _ = SEC.check_seq(st, 'alice', 10)
eq(c1, false, 'check_seq replay same')
local c2, _ = SEC.check_seq(st, 'alice', 11)
eq(c2, true, 'check_seq next ok')
local c3, m = SEC.check_seq(st, 'alice', 20)
eq(c3, true, 'check_seq gap')
eq(m, 'gap', 'check_seq gap reason')

-- independent peers
local rb, mb = SEC.add_seq(st, 'bob', 5)
eq(rb, true, 'bob first seq ok')
eq(mb, 'ok', 'bob first seq message')

-- ---- Results ----

print(string.format('\n%d passed, %d failed', passed, failed))
if failed > 0 then os.exit(1) end
